### ISSUE TITLE
cmor.set_grid_mapping has multiple problems.

### ISSUE THREAD

**Issue by JimBiardCics** [PCMDI/cmor#140]
A call to the cmor.set_grid_mapping function fails in pywrapper.py line 222 with the error

NameError: "global name 'punits' is not defined"

when the parameter_names argument to the function is a dictionary and contains an element that is a tuple or list of the form (value, units), because the punit and punits should be the same variable.

I fixed this problem by changing all instances of punits to punit, then found that there is another naming problem in the same function. The variables pnams and pnms should be the same variable. If you change all instances of pnms to pnams, the function successfully gets down to the call to the function _cmor.set_grid_mapping.

**Comment by dnadeau4** [comment]
Can you provide me with a test?

**Comment by JimBiardCics** [comment]
@dnadeau4 Inspection should be sufficient, but I can. I'll get you one shortly.

**Comment by dnadeau4** [comment]
@JimBiardCics I like test, I can add them to my test suite and make sure things are still working as they should before the release.

**Comment by JimBiardCics** [comment]
Here's a python script that creates a grid and sets a grid mapping on it, along with tables that it needs.

[SetGridMappingTest.zip](https://github.com/PCMDI/cmor/files/817681/SetGridMappingTest.zip)

To run it in the folder it is found in type:
`python testCmorGrid.py Tables/obs4MIPs_SImon.json cmor_config.json CMIP6_grids.json`

You can also use this script to check out what's up with the obs4MIPs_grids.json file.

**Comment by dnadeau4** [comment]
You are not using  [cmor.set_grid_mapping()](https://cmor.llnl.gov/mydoc_cmor3_api/#cmorsetgridmapping) properly.  
[SetGridMappingTest.zip](https://github.com/PCMDI/cmor/files/918143/SetGridMappingTest.zip)

```python
    mapnm = 'polar_stereographic'
    params = [ "standard_parallel",
               "latitude_of_projection_origin",
               "false_easting",
               "false_northing",
               "straight_vertical_longitude_from_pole",
               "scale_factor_at_projection_origin"]

    punits = ["degrees_north","degrees_north","meters","meters","degrees_east","" ]
    pvalues = [70.,90.,0.,0.,135.0, 1. ]

    cmor.set_grid_mapping(grid_id=grid_id, 
                      mapping_name='polar_stereographic',
                      parameter_names = params,
                      parameter_values = pvalues,
                      parameter_units = punits)

```

**Comment by doutriaux1** [comment]
@dnadeau4 @JimBiardCics can we close this then?

**Comment by dnadeau4** [comment]
@JimBiardCics should close it.

**Comment by JimBiardCics** [comment]
Per the API documentation in pywrapper.py:

```
def set_grid_mapping(grid_id,mapping_name,parameter_names,parameter_values=None,parameter_units=None):
    """Sets the grid mapping for CF convention
    Usage:
       set_grid_mapping(grid_id,mapping_name,parameter_names,parameter_values,parameter_units)
    Where:
       grid_id :: grid_id return by cmor.grid
       mapping_name     :: name of the mapping (see CF conventions)
       parameter_names  :: list of parameter names or dictionary with name as keys and values can be either [value,units] list/tuple or dictionary with "values"/"units" as keys
       parameter_values :: array/list of parameter values in the same order of parameter_names (ignored if parameter_names is ditcionary)
       parameter_units  :: array/list of parameter units  in the same order of parameter_names (ignored if parameter_names is ditcionary)
    """
```
It says that parameter_names can be a dictionary rather than doing the separate names, values, and units lists. I'm just using the API as it is advertised.

**Comment by dnadeau4** [comment]
@JimBiardCics thanks,  I can see the python wrapper code.   I'll fix it, before release 3.2.3

**Comment by dnadeau4** [comment]
Fixed:
https://github.com/PCMDI/cmor/pull/159
